### ISSUE TITLE
Math Notepad needs to become open source! 

### ISSUE THREAD

**Issue by cryptomanxxx** [josdejong/mathjs#1593]
All current mathematical software / computer algebra systems (cas) such as Maple, Mathematica and Matlab have close source code and they are proprietary. They only exist to make money. I refuse to use software from greedy and pathetic capitalists that are only interested in making money. Math notepad have a golden opportunity here to take the lead. Math notepad therefore needs to become open source and I am not the first person who have asked for this. People have asked for this for a long time but nothing happens. 

As a scientist I need a high quality open source code java script web editor where I can write my own JavaScript functions and do my own data analysis. I have tried my best to create one but it is hard when I dont have a programming background. https://github.com/cryptomanxxx/JavaScript_without_HTML 

It irritates me that I have to waste time to invent the wheel all over again when the knowledge how to create one already exist but only available to people with money. In a black and white world there is no grey. Either Math notepad is part of the problem (close source code and proprietary) or math notepad is part of the solution. Please make the right choice fast.

**Comment by harrysarson** [comment]
If I may answer this question on behalf of Jos, who has spent years contributing over 2000 commits to this open source mathematics library, has given up a lot of time to engage with folk using this open source library and has put huge amounts of effort into constantly improving this open source library largely based on requests from users.

Creating high quality mathematical software _is_ hard and the reason that companies like MathWorks exist and can charge customers for software such as matlab is that they provide a service that people are willing to pay for. The cost of reinventing the wheel is less than a license for Mathmatica.

Jos has no obligation to opensource any of his work. (I resist the urge to digress on a tangent about the obligation Jos does have to opensource any software he has written deriving from software with a copyleft license.) Many people, myself included, are grateful he has chosen to open source this high quality mathematics library. The library is high quality as a result of much _hard_ work on the parts of Jos and the other contributors. 

I am going to use my contributor rights to close this issue due to the confrontation tone and complete lack of any actional points.

**Comment by josdejong** [comment]
I'm always open for any constructive discussion, but I'm not going to comment on such a rude, negative, demanding tone.

**Comment by cryptomanxxx** [comment]
I apologize if the tone was hard but the fact is that there is war going on between people that want to profit from knowledge, control it and have closed science (universities, mathematical software companies and greedy scientific journal publishing companies) vs open science and free flow of knowledge. As an independent scientist (not spoiled by a wage from a university) that lives on an constrained income I have despite this fact always published my scientific papers open source and always paid my publication fees myself (because I believe in open science and free flow of knowledge) but I refuse to pay money to a company (mathematical software company) that is so obviously taking advantage of me by using knowledge to suppress and control me. 

We all use other peoples knowledge and work everyday and without knowledge sharing we would be still sitting in caves throwing shit at each other. Try building your own computer, grow your own food, slaughter your own animals, make your own medicines and generate your own eletricity and at the same time you do that try to do what you love and you will soon find out it is impossible because there is not enough time in a day! Knowledge should flow feely and certainty not be used to suppress and control people like a capitalist universities does. Since Jos, who has apparently contributed a lot of his time to the open source community (which we all appreciate), I think he would also appreciate and value the open source community and free flow of knowledge. Learning how to fish is more important for a man / woman than being given fish or no fish at all for that matter.

**Comment by josdejong** [comment]
Apologies accepted.

I'm glad you see a lot of potential in Math Notepad. Please note that I've open sourced the largest part of Math Notepad already in the form of this `mathjs` library. Just open sourcing Math Notepad will not solve any immediate issue. Running an open source project successfully is not as simple as you may think. It requires dedicated people putting a lot of effort (for free!) in the project.

If you're looking for open source solutions for maths and data analysis, there is a lot of good stuff out there: have a look at [Scilab](https://www.scilab.org/), [Octave](https://www.gnu.org/software/octave/), python [Jupyter](https://jupyter.org/) and [Pandas](https://pandas.pydata.org/), [RStudio](https://www.rstudio.com/), ...

**Comment by cryptomanxxx** [comment]
ok thx for the suggestions and links. I will check them out and see if they can add any value to my research / data analysis.

**Comment by cryptomanxxx** [comment]
ok, let me see if I can explain more pedagogically by using an example what kinds of problems not open sourcing the HTML for math notepad leads to for me and for others.  

Let say I write a java script function that I have stored locally in a text file on my computer. 
Now I want to run that function on your website http://mathnotepad.com/ 
There is no way for me to do that because only the JavaScript functions that can be run on your website math notepad are defined by you and located on your server.

For me this becomes a huge design flaw! Why should you decide and control what JavaScript functions I am allowed to write and run on math notepad? That is why I am saying that it is better to open source the html for math notepad and let everyone host that html file how ever they please. Locally or on a server. Then I (or anyone ells for that matter) can modify the HTML to include a link to any text file with user specified functions that can be called from Math notepad.

I have already wasted at least 4-5 years programming with a proprietary mathematical software (Maple) https://maplesoft.com/applications/Author.aspx?mid=87231 where the developers have 100% control over the programming environment. Since I dont want to pay that company money each year plus I am not that impressed with the software on my touch screen device my previous work have been wasted. Letting developers have 100% control over the programming environment is a huge risk that no users should accept! I dont want to repeat the same mistake with math notepad. 

If I am going to use math notepad (or anyone ells for that matter) then the HTML for math notepad must become open source so I (or anyone ells) can reference user specified java script functions and have 100% control. An alternative would be to allow users themself to write and run functions directly on your website math notepad but the HTML does not seem to be sophisticated enough for that to happen at the moment.

**Comment by harrysarson** [comment]
It does sound like numpy + Jupyter is exactly what you are searching for. I think you could have some luck exploring those.

**Comment by cryptomanxxx** [comment]
numpy is python and I have never liked the programming language python! So unfortunately you are wrong. I am very sensitive to programming language syntax. If I dont like the syntax in the first couple of days I will drop that language. The only programming languages I more or less like are: JavaScript, Maple, VBA and google aps script.

**Comment by cryptomanxxx** [comment]
The fact is that I have even seen selfish and anal developers that have inserted malicious code in to the open source code to make sure they have 100% control of code, in this case, a crypto currency wallet. If the developers did not like what people were doing with "their" code they could externally (a kill switch) make the code stop working. This is the kind of hostility we see from some developers that do not truely support open source code. The fact is that when developers spend too (?) much time on developing the code they become very possessive of it and some time very selfish which can have a very negative impact on users. I am not saying this will happen with math notepad but the fact is that if the you dont opensource the code I can promise you that someone will develop the same thing and open source it. Then all your work unfortunatly for you has been for nothing.

**Comment by cryptomanxxx** [comment]
Also I have had a look at the HTML of the expression parser of mathjs in the browser https://github.com/josdejong/mathjs-expression-parser 

What kind of ad hoc and un-ortodox src attribute is: 

<script src="../dist/mathjs-expression-parser.js"></script>

Does ... mean I dont give a crap if this work for you since it works for me?

A correct src attribute look like this:

<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script> 

There already exist a few red flags for me with this project to say the least.

**Comment by josdejong** [comment]
I will close this discussion, it's not constructive at all.

**Comment by ericman314** [comment]
Jos, I know I might regret posting in fear that it may encourage even more toxic discussion in the future, but I thought the least I could do is fork cryptoman's repo and alter it to use math.js as a processor. I hope you are not offended by this. 

@cryptomanxxx, I share your preference for open source software, and I rarely (if ever) pay for software. But in return I contribute to the open source community in a small way by donating my time and energy to moving projects like math.js forward. I appreciate that you fund your own research and publish in open access journals. I also understand your frustration with learning new programming languages and syntaxes. For most people, the only contribution they can reasonably make towards open source is just their gratitude, because it is that gratitude which motivates the rest of us to continue maintaining and developing these projects.

Unfortunately I think you have offended too many people here to get much more help with math.js, but I hope the little change I made to your repo can at least get you started: https://github.com/ericman314/JavaScript_without_HTML
